pkg/adapter/awssnssource/adapter.go

```diff
@@ -18,216 +18,155 @@ package awssnssource
-		return
-	}
-
-	a.logger.Debug("Request body: ", string(body))
-
-	switch data["Type"].(string) {
-	// If the message is about our subscription, call the confirmation endpoint.
-	// payload: https://docs.aws.amazon.com/sns/latest/dg/sns-message-and-json-formats.html#http-subscription-confirmation-json
-	case "SubscriptionConfirmation":
-		resp, err := a.snsClient.ConfirmSubscription(&sns.ConfirmSubscriptionInput{
-			TopicArn: aws.String(data["TopicArn"].(string)),
-			Token:    aws.String(data["Token"].(string)),
-		})
-		if err != nil {
-			a.logger.Errorw("Unable to confirm SNS subscription", zap.Error(err))
-			http.Error(rw, fmt.Sprint("Unable to confirm SNS subscription: ", err), http.StatusInternalServerError)
-			return
-		}
-
-		a.logger.Debug("Successfully confirmed SNS subscription: ", *resp)
-
-	// If the message is a notification, push the event
-	// payload: https://docs.aws.amazon.com/sns/latest/dg/sns-message-and-json-formats.html#http-notification-json
-	case "Notification":
-		event := cloudevents.NewEvent(cloudevents.VersionV1)
-		event.SetType(v1alpha1.AWSEventType(a.arn.Service, v1alpha1.AWSSNSGenericEventType))
-		event.SetSource(a.arn.String())
-		event.SetID(data["MessageId"].(string))
-
-		if subjectAttr, ok := data["Subject"]; ok {
-			event.SetSubject(subjectAttr.(string))
-		}
-
-		if err := event.SetData(cloudevents.ApplicationJSON, body); err != nil {
-			a.logger.Errorw("Failed to set event data", zap.Error(err))
-			http.Error(rw, fmt.Sprint("Failed to set event data: ", err), http.StatusInternalServerError)
-			return
+		if err := s.Shutdown(ctx); err != nil {
+			return fmt.Errorf("during server shutdown: %w", err)
 		}
 
-		if result := a.ceClient.Send(context.Background(), event); !cloudevents.IsACK(result) {
-			a.logger.Errorw("Failed to send CloudEvent", zap.Error(result))
-			http.Error(rw, fmt.Sprint("Failed to send CloudEvent: ", result), http.StatusInternalServerError)
-		}
+		return handleServerError(<-errCh)
 
-		a.logger.Debug("Successfully sent SNS notification: ", event)
+	case err := <-errCh:
+		return handleServerError(err)
 	}
 }
 
-func healthCheckHandler(w http.ResponseWriter, _ *http.Request) {
-	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
-	w.WriteHeader(http.StatusOK)
-	fmt.Fprintln(w, "OK")
+// ServeHTTP implements http.Handler.
+// Delegates incoming requests to the underlying router.
+func (a *adapter) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	a.router.ServeHTTP(w, r)
 }
 
 // RegisterHandlerFor implements MTAdapter.
 func (a *adapter) RegisterHandlerFor(ctx context.Context, src *v1alpha1.AWSSNSSource) error {
-	// TODO(antoineco): implement routing
+	snsCli, err := a.snsCg.Get(src)
+	if err != nil {
+		return fmt.Errorf("obtaining SNS client: %w", err)
+	}
+
+	h := handler.New(src, a.logger, a.ceClient, snsCli)
+
+	a.router.RegisterPath(routing.URLPath(src), h)
 	return nil
 }
```

Comment:
The aforementioned handler is generated using:
* a _CloudEvent client_, for sending events to their configured sink
* the _SNS client_ obtained from the getter and preconfigured with the region and credentials referenced in the source